File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-java/src/main/java/com/example/statemachine/simple/doument/persist/DocumentStateUpdater.java
package com.example.statemachine.simple.doument.persist;

import com.example.statemachine.simple.doument.DocumentEvent;
import com.example.statemachine.simple.doument.DocumentState;
import java.util.Objects;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.messaging.Message;
import org.springframework.statemachine.StateMachine;
import org.springframework.statemachine.recipes.persist.AbstractPersistStateMachineHandler.GenericPersistStateChangeListener;
import org.springframework.statemachine.state.State;
import org.springframework.statemachine.transition.Transition;

public class DocumentStateUpdater
  implements GenericPersistStateChangeListener<DocumentState, DocumentEvent> {

  private final JdbcTemplate jdbcTemplate;

  public DocumentStateUpdater(final JdbcTemplate jdbcTemplate) {
    this.jdbcTemplate = jdbcTemplate;
  }

  @Override
  public void onPersist(
    State<DocumentState, DocumentEvent> state, Message<DocumentEvent> message,
    Transition<DocumentState, DocumentEvent> transition,
    StateMachine<DocumentState, DocumentEvent> stateMachine) {

    if (Objects.nonNull(message) && message.getHeaders().containsKey("document")) {
      final var documentId = message.getHeaders().get("document", Integer.class);

      jdbcTemplate.update(
        "update documents set state = ? where id = ?",
        state.getId().toString(), documentId);
    }
  }
}

File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-kotlin/src/main/kotlin/com/example/statemachine/simple/document/persist/DocumentStateUpdater.kt
package com.example.statemachine.simple.document.persist

import com.example.statemachine.simple.document.DocumentEvent
import com.example.statemachine.simple.document.DocumentState
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.messaging.Message
import org.springframework.statemachine.StateMachine
import org.springframework.statemachine.recipes.persist.AbstractPersistStateMachineHandler
import org.springframework.statemachine.state.State
import org.springframework.statemachine.transition.Transition

class DocumentStateUpdater(private val jdbcTemplate: JdbcTemplate) :
  AbstractPersistStateMachineHandler.GenericPersistStateChangeListener<DocumentState?, DocumentEvent?> {


  override fun onPersist(
    state: State<DocumentState?, DocumentEvent?>?,
    message: Message<DocumentEvent?>?,
    transition: Transition<DocumentState?, DocumentEvent?>?,
    stateMachine: StateMachine<DocumentState?, DocumentEvent?>?
  ) {
    if (message!!.headers.containsKey("document")) {
      val documentId = message.headers.get("document", Integer::class.java)

      jdbcTemplate.update(
        "update documents set state = ? where id = ?",
        state!!.id.toString(), documentId
      )
    }
  }
}
